File: ts-backend/src/middleware/rate-limit.ts
/**
 * Per-IP token-bucket rate limiter middleware for the Bun+Hono server.
 *
 * Ports backend/middleware/rate_limit.py + backend/rate_limiter.py to TypeScript
 * with exact behavioral parity.
 *
 * Parameters (matching Python backend exactly):
 *   rate=1.0 tokens/second, burst=60.0 tokens
 *   cleanup_interval=60.0s, stale_after=600.0s
 *
 * IP source:
 *   Derived from the raw socket / Bun server's remoteAddr — the TRUE peer IP.
 *   We intentionally do NOT read X-Forwarded-For; it is spoofable (CWE-348).
 *   In Hono on Bun, the peer IP is obtained via c.env?.server (Bun.Server)
 *   requestIP() helper. Falls back to "unknown" — never XFF.
 *
 * Rate-limit exemptions (matching legacy api.py exemptions exactly):
 *   /health, /health/loop, /health/modules — exempt
 *   /metrics — exempt
 *
 * 429 response body matches Python _send_429 exactly:
 *   {"error": "rate limit exceeded", "retry_after": <int>}
 *   Headers: Content-Type: application/json, Retry-After: <int>,
 *            X-RateLimit-Remaining: 0
 *
 * Disable: set AF_RATE_LIMIT_DISABLED=1
 */

import type { Context, MiddlewareHandler, Next } from "hono";

// ---------------------------------------------------------------------------
// Paths exempt from rate limiting — must match Python backend exactly.
// ---------------------------------------------------------------------------
export const RATE_LIMIT_EXEMPT_PATHS: ReadonlySet<string> = new Set([
  "/health",
  "/health/loop",
  "/health/modules",
  "/metrics",
]);

// ---------------------------------------------------------------------------
// Token bucket — mirrors Python TokenBucket in backend/rate_limiter.py
// ---------------------------------------------------------------------------
export class TokenBucket {
  private readonly rate: number;
  private readonly burst: number;
  private tokens: number;
  private lastRefill: number; // performance.now() ms

  constructor(rate: number, burst: number) {
    this.rate = rate;
    this.burst = burst;
    this.tokens = burst; // new bucket starts full
    this.lastRefill = performance.now();
  }

  private refill(): void {
    const now = performance.now();
    const elapsed = (now - this.lastRefill) / 1000; // convert ms → seconds
    this.tokens = Math.min(this.burst, this.tokens + elapsed * this.rate);
    this.lastRefill = now;
  }

  consume(): boolean {
    this.refill();
    if (this.tokens >= 1.0) {
      this.tokens -= 1.0;
      return true;
    }
    return false;
  }

  tokensRemaining(): number {
    this.refill();
    return this.tokens;
  }

  retryAfter(): number {
    this.refill();
    if (this.tokens >= 1.0) return 0;
    const deficit = 1.0 - this.tokens;
    return deficit / this.rate;
  }

  get lastSeen(): number {
    return this.lastRefill;
  }
}

// ---------------------------------------------------------------------------
// Rate limiter — mirrors Python RateLimiter in backend/rate_limiter.py
// ---------------------------------------------------------------------------
export class RateLimiter {
  private readonly rate: number;
  private readonly burst: number;
  private readonly staleAfterMs: number;
  private readonly cleanupIntervalMs: number;
  private readonly buckets: Map<string, TokenBucket> = new Map();
  private lastCleanup: number; // performance.now() ms

  constructor(
    rate = 1.0,
    burst = 60.0,
    cleanupInterval = 60.0, // seconds
    staleAfter = 600.0 // seconds
  ) {
    this.rate = rate;
    this.burst = burst;
    this.staleAfterMs = staleAfter * 1000;
    this.cleanupIntervalMs = cleanupInterval * 1000;
    this.lastCleanup = performance.now();
  }

  private getOrCreate(ip: string): TokenBucket {
    let bucket = this.buckets.get(ip);
    if (!bucket) {
      bucket = new TokenBucket(this.rate, this.burst);
      this.buckets.set(ip, bucket);
    }
    return bucket;
  }

  private maybeCleanup(): void {
    const now = performance.now();
    if (now - this.lastCleanup < this.cleanupIntervalMs) return;
    const cutoff = now - this.staleAfterMs;
    for (const [ip, bucket] of this.buckets) {
      if (bucket.lastSeen < cutoff) this.buckets.delete(ip);
    }
    this.lastCleanup = now;
  }

  check(ip: string): [boolean, number] {
    this.maybeCleanup();
    const bucket = this.getOrCreate(ip);
    const allowed = bucket.consume();
    const remaining = bucket.tokensRemaining();
    return [allowed, remaining];
  }

  retryAfter(ip: string): number {
    const bucket = this.buckets.get(ip);
    if (!bucket) return 0;
    const secs = bucket.retryAfter();
    return Math.max(1, Math.floor(secs) + 1);
  }

  bucketCount(): number {
    return this.buckets.size;
  }
}

// ---------------------------------------------------------------------------
// Shared limiter instance — same parameters as Python api.py:2017-2019
// ---------------------------------------------------------------------------
export const sharedLimiter = new RateLimiter(1.0, 60.0, 60.0, 600.0);

// ---------------------------------------------------------------------------
// Peer IP extraction
// ---------------------------------------------------------------------------

/**
 * Extract the true peer IP from a Hono/Bun context.
 *
 * Bun exposes the server via c.env?.server (BunServer) which has a
 * requestIP(request) method returning { address, port, family }.
 * We NEVER read X-Forwarded-For — it is spoofable (CWE-348).
 */
export function getPeerIp(c: Context): string {
  try {
    const server = (
      c.env as
        | {
            server?: {
              requestIP?: (
                req: Request
              ) => { address: string } | null;
            };
          }
        | undefined
    )?.server;
    if (server?.requestIP) {
      const addr = server.requestIP(c.req.raw);
      if (addr?.address) return addr.address;
    }
  } catch {
    // Fall through to unknown
  }
  return "unknown";
}

// ---------------------------------------------------------------------------
// Hono middleware
// ---------------------------------------------------------------------------

export const rateLimitMiddleware: MiddlewareHandler = async (
  c: Context,
  next: Next
): Promise<void | Response> => {
  // Check disable flag
  if (process.env.AF_RATE_LIMIT_DISABLED === "1") {
    await next();
    return;
  }

  const path = new URL(c.req.url).pathname;

  // Exempt paths pass through unconditionally
  if (RATE_LIMIT_EXEMPT_PATHS.has(path)) {
    await next();
    return;
  }

  const ip = getPeerIp(c);
  const [allowed] = sharedLimiter.check(ip);

  if (!allowed) {
    const retryAfter = sharedLimiter.retryAfter(ip);
    return new Response(
      JSON.stringify({ error: "rate limit exceeded", retry_after: retryAfter }),
      {
        status: 429,
        headers: {
          "Content-Type": "application/json",
          "Retry-After": String(retryAfter),
          "X-RateLimit-Remaining": "0",
        },
      }
    );
  }

  await next();
};

File: backend/middleware/rate_limit.py
"""
Per-IP token-bucket rate limiter middleware for the FastAPI app.

Ports the legacy api.py _check_rate_limit / _send_429 behaviour to a
Starlette BaseHTTPMiddleware so the FastAPI app applies the same throttling
as the legacy ThreadingHTTPServer.

Parameters (matching legacy api.py exactly):
  rate=1.0 tokens/second, burst=60.0 tokens
  cleanup_interval=60.0s, stale_after=600.0s

IP source:
  ``request.client.host`` — the app strips X-Forwarded-* headers at the
  reverse-proxy layer, so ``client.host`` is the real peer IP.
  We intentionally do NOT read X-Forwarded-For; it is spoofable.

Rate-limit exemptions:
  /health, /health/loop, /health/modules — exempt in legacy (api.py:2421-2440)
  /metrics — exempt in legacy (sits before _check_rate_limit in api.py:2649)

429 response body matches legacy _send_429 exactly:
  {"error": "rate limit exceeded", "retry_after": <int>}
  Headers: Content-Type: application/json, Retry-After: <int>,
           X-RateLimit-Remaining: 0

SSE / WebSocket:
  The middleware gate fires on every request scope, so SSE/WS connections
  are gated on connection open — not on every keep-alive frame — which
  matches the legacy behaviour (legacy applies _check_rate_limit once per
  GET /stream/* or GET /ws request, before the streaming loop starts).

Disable at startup:
  Set AF_RATE_LIMIT_DISABLED=1 to pass all requests through (mirrors
  legacy --no-rate-limit flag), or call RateLimitMiddleware.enabled = False
  before adding the middleware (useful in tests).
"""

from __future__ import annotations

import json
import os

from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response

from backend.rate_limiter import RateLimiter

# ---------------------------------------------------------------------------
# Paths exempt from rate limiting — must match legacy api.py exemptions.
# /health, /health/loop, /health/modules are served before _check_rate_limit.
# /metrics is also served before the rate-limit gate in legacy (api.py:2649).
# ---------------------------------------------------------------------------
_EXEMPT_PATHS: frozenset[str] = frozenset(
    {
        "/health",
        "/health/loop",
        "/health/modules",
        "/metrics",
    }
)

# Shared limiter instance — same parameters as api.py:2017-2019.
_limiter: RateLimiter = RateLimiter(
    rate=1.0,
    burst=60.0,
    cleanup_interval=60.0,
    stale_after=600.0,
)


class RateLimitMiddleware(BaseHTTPMiddleware):
    """Per-IP token-bucket rate limiter.

    Returns HTTP 429 + Retry-After when the IP exhausts its token bucket.
    Exempt paths (health/metrics) pass through unconditionally.

    Set ``RateLimitMiddleware.enabled = False`` before instantiation (or set
    env var ``AF_RATE_LIMIT_DISABLED=1``) to disable for tests or the
    ``--no-rate-limit`` flag equivalent.
    """

    # Class-level toggle — mirrors legacy _Handler.enable_rate_limit.
    enabled: bool = True

    # Expose the shared limiter so tests can inspect bucket state.
    limiter: RateLimiter = _limiter

    async def dispatch(self, request: Request, call_next) -> Response:  # type: ignore[override]
        # Check env-var disable flag at dispatch time so it can be set after
        # app construction (e.g. in test fixtures).
        if not self.__class__.enabled or os.environ.get("AF_RATE_LIMIT_DISABLED") == "1":
            return await call_next(request)

        path = request.url.path
        if path in _EXEMPT_PATHS:
            return await call_next(request)

        ip = request.client.host if request.client else "unknown"
        allowed, _remaining = self.__class__.limiter.check(ip)

        if not allowed:
            retry_after = self.__class__.limiter.retry_after(ip)
            body = json.dumps(
                {"error": "rate limit exceeded", "retry_after": retry_after}
            ).encode()
            return Response(
                content=body,
                status_code=429,
                media_type="application/json",
                headers={
                    "Retry-After": str(retry_after),
                    "X-RateLimit-Remaining": "0",
                },
            )

        return await call_next(request)
